source/common/http/conn_pool_grid.cc

```diff
@@ -391,11 +397,19 @@ ConnectionPool::Cancellable* ConnectivityGrid::newStream(Http::ResponseDecoder&
   ConnectionPool::Instance* pool = getOrCreateHttp3Pool();
   Instance::StreamOptions overriding_options = options;
   bool delay_tcp_attempt = true;
+  bool delay_alternate_http3_attempt = true;
   if (shouldAttemptHttp3() && options.can_use_http3_) {
     if (getHttp3StatusTracker().hasHttp3FailedRecently()) {
       overriding_options.can_send_early_data_ = false;
       delay_tcp_attempt = false;
     }
+    if (http3_pool_ && http3_alternate_pool_ && !http3_pool_->hasActiveConnections() &&
+        http3_alternate_pool_->hasActiveConnections()) {
+      // If it looks like the main HTTP/3 pool is not functional and the
+      // alternate works, don't wait 300ms before attempting to use the
+      // alternate pool.
+      delay_alternate_http3_attempt = false;
+    }
```

Comment:
Does this mean we'll end up racing an H3 connection to the main pool with the alternate pool? If so, that seems like a bummer because we could potentially be trying lots of connection attempts to an IP we can't reach. In this case can we skip the main pool attempt? (Or maybe this doesn't quite work the way I think)

Reply:
yeah we'll race.  Would you be Ok with a TODO?  The other attempt should be canceled pretty quickly and it's a lot more complicated to only do the H3 alternate.  I think racing is better in this case than 300ms delay and I can't get the more complicated fix in before I take off.

Reply:
Hm. Will the QUIC handshake actually be cancelled, or will it keep running in the pool. If it's the former, I guess that's OK. But if it's the latter, that seems quite undesirable. I *guess* we could TODO this case. It just feels super ugly to potentially run a failing QUIC handshake on, say every request for a video chunk seems depressing. But, yeah, better than 300ms. Just not sure if this is better or worse than using TCP in this case?

Reply:
it'll be canceled as soon as the other succeeds.